bin/build_update_subfolders.py

```diff
@@ -0,0 +1,110 @@
+    """
+    Used to check if the submitter is new or already has an existing folder.
+
+    This function uses ``synapseutils.walk`` to crawl through the level 1 subfolders
+    (i.e. The folders 1 level below the Parent-Folder, that are named after the submitter's synapse ID) to
+    verify whether or not a submitter already has an existing folder created.
+    """
+
+    submitter_subfolders = []
+    # Setting ``include_types`` to only Folders so that Files are excluded, and walk is faster.
+    for _, level1_subfolders, _ in synapseutils.walk(syn=syn, synId=parent_folder_id,
+                                                     includeTypes=["folder"]
+                                                     ):
+        for subfolder in level1_subfolders:
+            # ``folder`` is a tuple structured as follows: (Folder name, synID). Let's just get the Folder name.
+            submitter_subfolders.append(subfolder[0])
+
+    if submitter_id in submitter_subfolders:
+        return False
+
+    return True
+
+
+def build_subfolder(syn, folder_name, parent_folder):
+    """
+    """
+
+    # Create Folder object
+    subfolder = synapseclient.Folder(
+        name=folder_name, parent=parent_folder
+    )
+    subfolder = syn.store(obj=subfolder)
+
+    return subfolder
+
+
+# TODO: https://sagebionetworks.jira.com/browse/IBCDPE-809
+# def update_subfolders():
+#     """"""
+
+
+def update_permissions(syn, subfolder, project_folder_id, access_type=["CREATE", "READ", "DOWNLOAD", "UPDATE", "DELETE"]):
+    """Updates the permissions (local share settings) of the given Folder/File"""
+    organizers_id = syn.get(project_folder_id, downloadFile=False).createdBy
+    syn.setPermissions(
+        subfolder, principalId=organizers_id, accessType=access_type
+        )
+
+
+def get_parent_and_project_folder_id(syn, parent_folder, project_name):
+    """"""
+    project_id = syn.findEntityId(name=project_name)
+    parent_folder_id = syn.findEntityId(name=parent_folder, parent=project_id)
+
+    return parent_folder_id, project_id
+
+
+def build_update_subfolders(
+        project_name, submission_id, build_or_update, subfolders=["workflow_logs", "predictions"],
+        only_admins="predictions", parent_folder="Logs"
+        ):
+    """
+    This function can either build a new set of log subfolders under 
+    a participant folder, or update an existing participant folder with
+    new ancilliary files.
+
+    The Folder structure is as follows:
+    >Parent-Folder/
+    >>Level 1 Subfolder (Submitter-Folder)/
+    >>>Level 2 Subfolder/
+    >>>> ...
+    """
+    # Establish access to the Synapse API
+    syn = synapseclient.login()
+
+    parent_folder_id, project_id = get_parent_and_project_folder_id(syn, parent_folder, project_name)
+    submitter_id = send_email.get_participant_id(syn, submission_id)[0]
+
+    if build_or_update == "build" and is_new_submitter(syn, submitter_id, parent_folder_id):
```

Comment:
I wonder - do you think we'll absolutely need the `is_new_submitter` check here?

I was going suggest forgoing the check and doing a try-catch with `build_subfolder()` instead, but after fact-checking myself, I found that creating a new Folder when one already exists won't throw any errors.  It also doesn't look like it'll reset the child entities either.  For example:

```python
>>> # Create new folder
>>> folder = synapseclient.Folder("Test", parentId="syn20682803")
>>> folder = syn.store(folder)
>>> folder.id
'syn53470398'
>>> 
>>> # Add file to folder
>>> file = synapseclient.File("Desktop/foo.txt", parentId=folder.id)
>>> file = syn.store(file)
>>> 
>>> # Check folder contents
>>> [f.get('name') for f in syn.getChildren(folder.id)]
['foo.txt']
>>> 
>>> # Attempt to create "Test" folder again
>>> folder = synapseclient.Folder("Test", parentId="syn20682803")
>>> folder = syn.store(folder)
>>> folder.id
'syn53470398'
>>> 
>>> # Check folder contents again
>>> [f.get('name') for f in syn.getChildren(folder.id)]
['foo.txt']
```

But then again... it's probably good practice to only "create" if it's needed..... 🙃

Reply:
Hi @vpchung, good point, that was the same thing I ran into - it seems like the Folder gets recreated without any errors and without affecting the child entities, so the try-catch wouldn't work. I need to figure out what's happening under the hood of synpy to see why this is.

As for this code, I chose to only create when needed to be cautious in case re-creating the Folder each time leads to some unforeseen consequences. It will take up more time to run `is_new_submission` as more submissions pile in, so if folks who know more about Synapse Folder entities don't see any problems with re-creating the Folders each time, I can do away with this func. cc @thomasyu888

Reply:
> need to figure out what's happening under the hood of synpy to see why this is.

hah, is it a bug or a feature? 😬 

>  if folks who know more about Synapse Folder entities don't see any problems with re-creating the Folders each time, I can do away with this func

To help us proceed, I decided to look at what the current orchestrator is doing.  If I understand the code correctly, it performs a [try-catch when creating submission folders](https://github.com/Sage-Bionetworks/SynapseWorkflowOrchestrator/blob/b45a318bbbef4923166603a4a266c6f9ed8cf714/src/main/java/org/sagebionetworks/Archiver.java#L169), so maybe the recreation of Folders is okay...?

Definitely confirm with @thomasyu888, haha.

Reply:
Recreating folders is ok, it does change the modified on date, but not the created on date which is what we probably care about more.  I think we can do away with this function which will save us time in the long run.

Reply:
That's fair. Thanks for investigating @vpchung , thanks @thomasyu888 ! Will retire this func in the next commit.